feat: universal provider

ganchoradkov opened the pull request:
# Description
This draft PR is the initial implementation of the `universal provider` for walletconnect v2 protocol. 


## Usage

```typescript
import { ethers } from "ethers";
import UniversalProvider from "@walletconnect/universal-provider";

//  Initialize the provider
const provider = await UniversalProvider.init({
  logger: "info",
  relayUrl: "ws://<relay-url>",
  projectId: "12345678",
  metadata: {
    name: "React App",
    description: "React App for WalletConnect",
    url: "https://walletconnect.com/",
    icons: ["https://avatars.githubusercontent.com/u/37784886"],
  },
  client: undefined, // optional instance of @walletconnect/sign-client
});

//  create sub providers for each namespace/chain
await provider.connect({
    namespaces: {
      eip155: {
        methods: [
          "eth_sendTransaction",
          "eth_signTransaction",
          "eth_sign",
          "personal_sign",
          "eth_signTypedData",
        ],
        chains: ["eip155:80001"],
        events: ["chainChanged", "accountsChanged"],
        rpcMap: {
          80001: "https://rpc-mumbai.matic.today",
        },
      },
      pairingTopic: "<123...topic>", // optional topic to connect to
      skipPairing: false, // optional to skip pairing ( later it can be resumed by invoking .pair())
    },
  });

  //  Create Web3 Provider
  const web3Provider = new ethers.providers.Web3Provider(provider);
```

## Events

```typescript
// Subscribe for pairing URI
provider.on("display_uri", (uri) => {
  console.log(uri);
});

// Subscribe to session ping
provider.on("session_ping", ({ id, topic }) => {
  console.log(id, topic);
});

// Subscribe to session event
provider.on("session_event", ({ event, chainId }) => {
  console.log(event, chainId);
});

// Subscribe to session update
provider.on("session_update", ({ topic, params }) => {
  console.log(topic, params);
});

// Subscribe to session delete
provider.on("session_delete", ({ id, topic }) => {
  console.log(id, topic);
});
```

## Provider Methods

```typescript
interface RequestArguments {
  method: string;
  params?: any[] | undefined;
}

// Send JSON RPC requests

/**
 * @param payload
 * @param chain - optionally specify which chain should handle this request
 * in the format `<namespace>:<chainId>` e.g. `eip155:1`
 */
const result = await provider.request(payload: RequestArguments, chain: string | undefined);
```


<!-- If valid for smoke test on feature add screenshots -->

## Due Dilligence

* [ ] Breaking change
* [x] Requires a documentation update
* [ ] Requires a e2e/integration test update

ganchoradkov committed:
feat: universal provider with eip155 as initial namespace

ganchoradkov committed:
feat: adds type to sendAsync args param

ganchoradkov committed:
feat: adds readme

ganchoradkov committed:
feat: adds return type

ganchoradkov committed:
chore: rm comment

ganchoradkov committed:
feat: add dispay_uri event to readme

ganchoradkov committed:
refactor: update readme

xzilja commented on providers/universal-provider/package.json:
```diff
@@ -0,0 +1,52 @@
+    "lint": "eslint -c '../../.eslintrc' --fix './src/**/*.ts'",
+    "prettier": "prettier --check '{src,test}/**/*.{js,ts,jsx,tsx}'"
+  },
+  "dependencies": {
```
I think some of these dependencies are only needed for testing purposes, correct? If that is the case, could you move relevant ones to `devDependencies` so they don't reflect as deps in npm registry. (re install them to update pakage lock file as well)

ganchoradkov replied:
> Looking goood fire Will need to wait for tests to get implemented for final approval. One question, what's the idea behind `subProviders` folder? Is it to simplify implementation even further so that we can create those small chain configs ourselves?

Think of `subProviders` as controllers. Each namespace will have its own implementation where logic can't be shared between chains.

xzilja commented on providers/universal-provider/README.md:
```diff
@@ -0,0 +1,96 @@
+    url: "https://walletconnect.com/",
+    icons: ["https://avatars.githubusercontent.com/u/37784886"],
+  },
+  client: undefined, // optional instance of @walletconnect/sign-client
```
What are the circumstances where optional instance of client might be needed?

ganchoradkov committed:
chore: rm yarn-lock

xzilja commented on providers/universal-provider/test/index.spec.ts:
```diff
@@ -0,0 +1,329 @@
+import { describe, it } from "vitest";
+// import Web3 from "web3";
```
Are tests coming latter on, or why are these commented out?

ganchoradkov replied:
Tests are on the way 🚄

xzilja reviewed (commented):
Looking goood 🔥 Will need to wait for tests to get implemented for final approval. One question, what's the idea behind `subProviders` folder? Is it to simplify implementation even further so that we can create those small chain configs ourselves?

arein commented on providers/universal-provider/README.md:
```diff
@@ -0,0 +1,96 @@
+import UniversalProvider from "@walletconnect/universal-provider";
+
+//  Initialize the provider
+const provider = await UniversalProvider.init({
```
nit: why provide the rpc in `connect` not here?

ganchoradkov replied:
the idea for the rpc config being in the `connect` fx and not in the `init` was that in some use-cases you might not know which chains will be used at the time of initialization

arein commented on providers/universal-provider/src/types/providers.ts:
```diff
@@ -0,0 +1,36 @@
+  NamespaceConfig,
+} from ".";
+
+export interface ISubProvider {
```
```suggestion
export interface IProvider {
```

arein reviewed (commented):
Awesome! minot nitpicks - Some tests and then LGTM

ganchoradkov committed:
refactor: renames subProviders to provider

ganchoradkov committed:
chore: add to previous commit

ganchoradkov committed:
refactor: renames type ISubProvider to IProvider

ganchoradkov committed:
refactor: refactor: renames type ISubProvider to IProvider

ganchoradkov committed:
feat: universal provider tests

ganchoradkov committed:
Merge branch 'v2.0' into gancho/feat/universal-provider

ganchoradkov committed:
chore: rm comment

xzilja commented on providers/universal-provider/test/index.spec.ts:
```diff
@@ -0,0 +1,237 @@
+        resolve();
+      }),
+    ]);
+    // expect provider to be disconnected
```
are these left over comments, or something you want to add in the future?

xzilja reviewed (approved):
Looks good, now that the tests are in. I think we can merge this (perhaps after you address those `FIXME` comments?). Would be good to make a release of this as well and perhaps create a pr to wagmi to add as v2 provider.

ganchoradkov committed:
feat: adds default chain switch & test for it

ganchoradkov committed:
chore: rm comment & .only

ganchoradkov committed:
refactor: rm signTransaction with ethers as its not supported
https://github.com/ethers-io/ethers.js/blob/master/packages/providers/src.ts/json-rpc-provider.ts#L270

ganchoradkov committed:
feat: revert to default chain after test

ganchoradkov committed:
refactor: more assets

ganchoradkov committed:
feat: export types from universal-provider

ganchoradkov committed:
chore: bump versions

ganchoradkov committed:
feat: adds `wallet_switchEthereumChain` handling

ganchoradkov committed:
chore: package-lock

ganchoradkov committed:
feat: adds uri type & sets session type

ganchoradkov committed:
fix: returns null on `wallet_switchEthereumChain`

xzilja commented on providers/universal-provider/README.md:
```diff
@@ -0,0 +1,112 @@
+//  Initialize the provider
+const provider = await UniversalProvider.init({
+  logger: "info",
+  relayUrl: "ws://<relay-url>",
```
Should this be wss?

xzilja commented on providers/universal-provider/package.json:
```diff
@@ -0,0 +1,52 @@
+    "prettier": "prettier --check '{src,test}/**/*.{js,ts,jsx,tsx}'"
+  },
+  "dependencies": {
+    "@walletconnect/jsonrpc-http-connection": "1.0.3",
```
Before release, could you double check that these are all up to date

xzilja commented on providers/universal-provider/src/UniversalProvider.ts:
```diff
@@ -0,0 +1,270 @@
+        case "cosmos":
+          //TODO:
+          break;
+        case "polygon":
```
I think polygon is handled by Eip155Provider as well, correct?

xzilja commented on providers/universal-provider/src/constants/values.ts:
```diff
@@ -0,0 +1,5 @@
+export const LOGGER = "error";
+
+export const RELAY_URL = "ws://relay.walletconnect.com";
```
```suggestion
export const RELAY_URL = "wss://relay.walletconnect.com";
```

xzilja reviewed (approved):
Super nice! This will unlock so many cool integrations! Left few minor comments and in general if possible, could you look over typescript `any` and add appropriate types where possible, for nicer end user experience.

arein commented on providers/universal-provider/README.md:
```diff
@@ -0,0 +1,112 @@
+      chains: ["eip155:80001"],
+      events: ["chainChanged", "accountsChanged"],
+      rpcMap: {
+        80001: "https://rpc-mumbai.matic.today",
```
```suggestion
        80001: "https://rpc.walletconnect.com?chainId=eip155:80001&projectId=<your walletconnect project id>",
```

ganchoradkov committed:
Merge branch 'v2.0' into gancho/feat/universal-provider

ganchoradkov committed:
updates readme to use wc rpc provder

Co-authored-by: Derek <alexanderderekrein@gmail.com>

ganchoradkov committed:
ws -> wss

Co-authored-by: Ilja <IljaDaderko@users.noreply.github.com>

ganchoradkov committed:
refactor: rm polygon as its handled by eip155

ganchoradkov committed:
chore: updates dependencies

bkrem commented on providers/universal-provider/package.json:
```diff
@@ -0,0 +1,53 @@
+    "build:types": "tsc",
+    "build:source": "rollup --config rollup.config.js",
+    "build": "npm run build:pre; npm run build:source; npm run build:types",
+    "test:pre": "rm -rf ./test/test.db",
```
Think this can be removed since it's a side effect for running tests in sign-client specifically

bkrem commented on providers/universal-provider/src/UniversalProvider.ts:
```diff
@@ -0,0 +1,267 @@
+import { RELAY_URL, LOGGER, STORAGE } from "./constants";
+
+export class UniversalProvider implements IUniversalProvider {
+  public client!: SignClient;
```
How come we need the non-null assertion operator here but didn't seem to previously in SignClient and other similar classes?

ganchoradkov replied:
Because the client is not initialized in the constructor and thus requiring adding `| undefinied`. 
This makes usage later on annoying due to constantly having to add checks such as `client?.`

Thought this solution is fine as attempting to use the client without it being initialized results in an exception such as
https://github.com/WalletConnect/walletconnect-monorepo/blob/gancho/feat/universal-provider/providers/universal-provider/src/UniversalProvider.ts#L180

bkrem replied:
Yeah makes total sense 👍 

I'm more surprised we didn't run into this before in e.g. SignClient since the pattern is the same, but it seems like the way the type referencing there refers back to the abstract class every time (i.e. `public session: ISignClient["session"]`) doesn't evoke this issue in the TS compiler

bkrem commented on providers/universal-provider/src/constants/values.ts:
```diff
@@ -0,0 +1,5 @@
+
+export const RELAY_URL = "wss://relay.walletconnect.com";
+
+export const STORAGE = "@universal_provider";
```
Would it make sense to use a standardised format for the storage key prefixing like we do for Core, Sign etc? https://github.com/WalletConnect/walletconnect-monorepo/blob/v2.0/packages/core/src/constants/core.ts#L5

Also in terms of future-proofing/versioning the storage for possible breaking changes

ganchoradkov committed:
chore: lint

ganchoradkov committed:
refactor: rm console.log

ganchoradkov committed:
chore: no-case-declarations

ganchoradkov committed:
refactor: Expected '===' and instead saw '=='

ganchoradkov committed:
refactor: uses standardized storage prefix

ganchoradkov committed:
chore: rm unsed test commands

bkrem commented on providers/universal-provider/src/UniversalProvider.ts:
```diff
@@ -0,0 +1,267 @@
+
+  private createProviders(): void {
+    if (!this.client) {
+      throw new Error("Client not initialized");
```
Consistency nitpick:

```suggestion
      throw new Error("Sign Client not initialized");
```

bkrem commented on providers/universal-provider/src/UniversalProvider.ts:
```diff
@@ -0,0 +1,267 @@
+      switch (namespace) {
+        case "eip155":
+          this.rpcProviders[namespace] = new Eip155Provider({
+            // @ts-ignore
```
Still needed?

bkrem commented on providers/universal-provider/src/UniversalProvider.ts:
```diff
@@ -0,0 +1,267 @@
+
+  private registerEventListeners(): void {
+    if (typeof this.client === "undefined") {
+      throw new Error("WalletConnect is not initialized");
```
Consistency nitpick:

```suggestion
      throw new Error("Sign Client is not initialized");
```

bkrem commented:
Amazing work on this @ganchoradkov 💯

ganchoradkov committed:
Update providers/universal-provider/src/UniversalProvider.ts

Co-authored-by: Ben Kremer <ben@walletconnect.com>

ganchoradkov committed:
Update providers/universal-provider/src/UniversalProvider.ts

Co-authored-by: Ben Kremer <ben@walletconnect.com>

ganchoradkov committed:
chore: rm ts-ignore

ganchoradkov committed:
Merge branch 'v2.0' into gancho/feat/universal-provider

ganchoradkov committed:
Merge branch 'v2.0' into gancho/feat/universal-provider

ganchoradkov committed:
Merge branch 'v2.0' into gancho/feat/universal-provider
